library/init/meta/interactive.lean

```diff
@@ -6,8 +6,11 @@ Authors: Leonardo de Moura
 prelude
 import init.meta.tactic init.meta.rewrite_tactic init.meta.simp_tactic
 import init.meta.smt.congruence_closure init.category.combinators
+import init.meta.lean_parser
 
 namespace interactive
+@[reducible] meta def parse {α : Type} (p : lean_parser α) : Type := α
```

Comment:
> I would hate to give up the strong typing.

While strong typing might just work out for the custom tactic syntax, it will definitely be limiting for the parser extensions.  We also want parsers where the type of the result is not statically: think of a parser for base64 strings that constructs a `vector byte n`--we cannot statically determine the size and hence the type of the inserted expression.

I think that the not-strongly-typed representation of expressions we use in the tactic monad works really well.  It is simple, effective, and flexible.  In the superposition prover I construct lots of terms by hand (with binders as well), and I haven't really had any big problems.  Most of that would not be (easily) possible with strongly-typed expressions.

If we really want to, we can always bolt on (incorrectly-phased) types to `expr` to catch errors in the common cases:
```lean
meta inductive {u} pexpr_for (α : Type u) : Type
| mk : pexpr → pexpr_for

meta constant mk_list_pexpr {α} : list (pexpr_for α) → pexpr_for (list α)
meta constant mk_tuple_pexpr {α β} : pexpr_for α → pexpr_for β → pexpr_for (α × β)

meta example : pexpr_for (ℕ × list ℕ) := mk_list_pexpr []
-- static type error!
```

(BTW: I tried to use `meta def mk_list_pexpr := sorry` first, but then the example fails because sorry is noncomputable.  Should we make sorry computable?)

@leodemoura Why do you think we'd get nasty error messages?  These would be the same error messages the elaborator would return if you entered the term directly in Lean syntax.